### PR TITLE
Feature/batch run calcrule

### PR DESCRIPTION
**Pull request by delcroip**
pseudo code (not tested, unlikely to run) for new batchrun approach

### CODE REVIEW

File: claim_batch/services.py [hunk 1]
```diff
@@ -381,10 +429,102 @@ def _get_capitation_region_and_district(location_id):
     return region_id, district_id
 
 
-
 def do_process_batch(audit_user_id, location_id, period, year):
     processed_ids = set()  # As we update claims, we add the claims not in relative pricing and then update the status
     logger.debug("do_process_batch location %s for %s/%s", location_id, period, year)
+    batch_run = new BatchRun
```
Comment (delcroip) [inline claim_batch/services.py:435]: This is the pseudo code

**Comment by dragos-dobre** [comment]
@delcroip Merge to main is done in purpose? I would expect to have the merge to develop.

**Comment by delcroip** [comment]
@dragos-dobre  fixed but the idea is not to merge it any way, it is only some pseudo code so far

File: claim_batch/services.py [hunk 2]
```diff
@@ -431,6 +432,109 @@ def _get_capitation_region_and_district(location_id):
...
+                    remunerated_amount += service.remunerated_amount
+                for item in claim.items:
+                    remunerated_amount += item.remunerated_amount
+            # 5 Generate BatchPayment per product (Ideally per pool but the notion doesn't exist yet)
+            # 5.1 filter a calculation valid for batchRun with context BatchPayment (got via 0.2)
+            if work_data.paymentplans:
+                for paymentplan in work_data.paymentplans:
+                    if paymentplan.calculation.active_for_object(batch_run, 'BatchPayment'):
+                        # 54.2 Execute the converter per product/batch run/claim (not claims)
+                        paymentplan.calculation.calculate(work_data, 'BatchPayment'))
```
Comment (sniedzielski) [inline claim_batch/services.py:530]: @delcroip I think we miss the part with creating BatchRun object after execution of batch run. 
like 
```
    from core.utils import TimeUtils
    created_run = BatchRun.objects.create(location_id=location_id, run_year=year, run_month=period,
                                          run_date=TimeUtils.now(), audit_user_id=audit_user_id,
                                          validity_from=TimeUtils.now())
    logger.debug("do_process_batch created run: %s", created_run.id)
```

File: claim_batch/services.py [hunk 3]
```diff
@@ -431,6 +432,109 @@ def _get_capitation_region_and_district(location_id):
...
+
+    # 0 prepare the batch run :  does it really make sense per location ? (Ideally per pool but the notion doesn't exist yet)
+    # 0.1 get all product concerned, all product that have are configured for the location
+    _, days_in_month = calendar.monthrange(year, period)
+    end_date = datetime.date(year, period, days_in_month)
+    period_quarter = period - 2 if period|divisibleby:4 else 0
+    period_sem = period - 5 if period|divisibleby:6 else 0
+
+    products = Product.objects\
+            .filter(validity_from__lte = start_date)\
```
Comment (sniedzielski) [inline claim_batch/services.py:445]: @delcroip 'start_date' is not defined in that scope of code. Could be 
    ```
_, days_in_month = calendar.monthrange(year, period)
    start_date = datetime.date(year, period, 1)
    end_date = datetime.date(year, period, days_in_month)
    period_quarter = period - 2 if period|4 else 0
    period_sem = period - 5 if period|6 else 0
```
?

File: claim_batch/services.py [hunk 4]
```diff
@@ -431,6 +432,109 @@ def _get_capitation_region_and_district(location_id):
...
+            if product.period_rel_prices = 'Y' or product.period_rel_prices_ip = 'Y' or product.period_rel_prices_op = 'Y':
+                start_date = datetime.date(year, 1, 1)
+            else if product.period_rel_prices = 'S' or product.period_rel_prices_ip = 'S' or product.period_rel_prices_op = 'S':    
+                start_date = datetime.date(year, period_sem, 1)
+            else if product.period_rel_prices = 'Q' or product.period_rel_prices_ip = 'Q' or product.period_rel_prices_op = 'Q':    
+                start_date = datetime.date(year, period_quarter, 1)
+            else:
+                start_date = datetime.date(year, period, 1)
+
+            work_data.product = product
```
Comment (sniedzielski) [inline claim_batch/services.py:464]: @delcroip What does mean 'work_data'? i see this is a object that cointains some fields like product, payment plans etc? 
Should be somewhere a definition of representation of this object?

File: claim_batch/services.py [hunk 5]
```diff
@@ -431,6 +432,109 @@ def _get_capitation_region_and_district(location_id):
...
+            # 1.2 get all the payment plan per product
+            work_data.paymentplans = PaymentPlan.objects\
+                .filter(valid_to__lte = start_date)\
+                .filter(valid_from__gte = end_date)\
+                .filter(product = product)
+            # 1.3 retrive all the Item & service per product
+            work_data.items  = ClaimItem.objects\
+                .filter(validity_from__lte = start_date)\
```
Comment (sniedzielski) [inline claim_batch/services.py:472]: @delcroip Is this statement correct with 'validity_from__lte' and validity_from__gte'? The another ones looks fine. But I have concerns about this one.

File: claim_batch/services.py (hunk 3 above)
Comment (delcroip) [reply claim_batch/services.py:445]: I was sure it was there ... Yes do it

File: claim_batch/services.py (hunk 4 above)
Comment (delcroip) [reply claim_batch/services.py:464]: I don't think it is required, python should work with this kind of object without class (some fields could be defined but other will be created on the fly by calcules)

File: claim_batch/services.py (hunk 5 above)
Comment (delcroip) [reply claim_batch/services.py:473]: you are right use validity_from__gte but actually it this the processed date of the claim that should be used remove the validity_from altogether

Comment (delcroip) [reply claim_batch/services.py:473]: for item, service contribution and claims

File: claim_batch/services.py (hunk 3 above)
Comment (sniedzielski) [reply claim_batch/services.py:445]: ok

File: claim_batch/services.py [hunk 6]
```diff
@@ -532,6 +542,94 @@ def do_process_batch(audit_user_id, location_id, period, year):
...
+# update the service and item valuated amount 
+def claim_batch_valuation(work_data, start_date, end_date):
+    allocated_contributions = get_allocated_premium(work_data.contributions, start_date, end_date)
+    # Sum up all item and service amount
+    value_hospital = 0
+    value_non_hospital = 0
+    periode_type, period_id = get_period(start_date, end_date); # M/Q/Y , 1-12/1-4/1
+    # ensure that the product need to be valuated for the given period
+    if period_rel_prices  == periode_type:
+        period_rate = get_relative_price_rate(work_data.product, 'B' periode_type, period_id) # migh be added in product service
```
Comment (sniedzielski) [inline claim_batch/services.py:566]: @delcroip Is this method somewhere defined or do we need to implement this?

Comment (sniedzielski) [reply claim_batch/services.py:566]: or is this the same as for 'create_index' (empty method) like below?

File: claim_batch/services.py [hunk 7]
```diff
@@ -532,6 +542,94 @@ def do_process_batch(audit_user_id, location_id, period, year):
...
 
 
+# Calculate allcated contributions
+def get_allocated_premium(premiums, start_date, end_date):
+    # go trough the contribution and find the allocated contribution 
+    allocated_premiums = 0
+    for premiums in premium:
+        allocation_start = max(policy.effective_date, start_date)
+        allocation_stop = min(start_date, policy.expiry_date)
+        policy_duration = policy.expiry_date(  - policy.effective_date
```
Comment (sniedzielski) [inline claim_batch/services.py:552]: @delcroip Is this done om purpose? 'policy_duration = policy.expiry_date(  - policy.effective_date '

File: claim_batch/services.py (hunk 6 above)
Comment (sniedzielski) [reply claim_batch/services.py:566]: Is this signature for that method is ok? 
```
# might be added in product service
def get_relative_price_rate(product, index_type, period_type, period_id):
    pass #just create the row in db
```

File: claim_batch/services.py (hunk 5 above)
Comment (sniedzielski) [reply claim_batch/services.py:473]: ok, noted.

File: claim_batch/services.py [hunk 8]
```diff
@@ -431,6 +432,207 @@ def _get_capitation_region_and_district(location_id):
 def do_process_batch(audit_user_id, location_id, period, year):
     processed_ids = set()  # As we update claims, we add the claims not in relative pricing and then update the status
```
Comment (sniedzielski) [inline claim_batch/services.py:433]: @delcroip Is this part going to be used in new batch run design? In old method this is used.

File: claim_batch/services.py [hunk 9]
```diff
@@ -431,6 +432,207 @@ def _get_capitation_region_and_district(location_id):
...
+                .filter(claim_pocess_stamp_gte = end_date  )\
+                .filter(Q(items__product_in = products) & Q(items__legacy_id__isnull = True) | Q(services__product_in = products) & Q(services__legacy_id__isnull = True))\
+                .prefetch_related(Prefetch('items', queryset=ClaimItem.objects.filter(legacy_id__isnull=True)))\
+                .prefetch_related(Prefetch('services', queryset=ClaimService.objects.filter(legacy_id__isnull=True)))
+            for claim in claims:
+                remunerated_amount = 0
+                for service in claim.services:
+                    remunerated_amount += service.remunerated_amount
+                for item in claim.items:
+                    remunerated_amount += item.remunerated_amount
```
Comment (sniedzielski) [inline claim_batch/services.py:533]: @delcroip Where this 'remunerated_amount' is used later? Can you give me an example?

Comment (sniedzielski) [reply claim_batch/services.py:533]: I see this comment '# 4 update the claim Total amounts if all Item and services got "valuated"'
I assume we need to add such line to update 'valuated' in Claim object, don't we (see below)?
` claim.valuated = remunerated_amount `

File: claim_batch/services.py [hunk 10]
```diff
@@ -468,43 +468,53 @@ def do_process_batch(audit_user_id, location_id, period, year):
...
                 .filter(validity_to__isnull = True)\
-                .filter(product = product)
+                .filter(product = product)\
+                .select_related('claim__health_facility')\
+                .orderby('claim__health_facility').orderby('claim_id')
             # 1.4 retrive all contributions per product // process_stamp should be the creation date
             work_data.contributions  = Premium.objects\
                 .filter(validity_from__lte = start_date)\
                 .filter(validity_from__gte = end_date)\
                 .filter(pocess_stamp_lte = end_date  )\
```
Comment (sniedzielski) [inline claim_batch/services.py:494]: @delcroip Does this field exist in 'Premium' object. I cannot see such name of field in this model.

Comment (delcroip) [reply claim_batch/services.py:494]: you are right pocess_stamp must be added on the database and the date when the Premium is saved in the database (managed by the SQL server)

File: claim_batch/services.py (hunk 9 above)
Comment (delcroip) [reply claim_batch/services.py:533]: no, it is a fields of the claim (<> than the claim.valuated), I just forget the save()

File: claim_batch/services.py (hunk 10 above)
Comment (sniedzielski) [reply claim_batch/services.py:494]: ok noted

**Pull request merged**